tests/expose_test.go

```diff
@@ -0,0 +1,384 @@
+				&expect.BExp{R: "0"},
+			}, 60*time.Second)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		Context("Expose NodePort UDP service", func() {
+			const servicePort = "27018"
+			const serviceName = "node-port-udp-vm"
+			const nodePort = "30018"
+
+			It("Should expose a NodePort service on a VM", func() {
+				virtctl := tests.NewRepeatableVirtctlCommand(expose.COMMAND_EXPOSE, "virtualmachine", "--namespace",
+					vm.Namespace, vm.Name, "--port", servicePort, "--name", serviceName, "--target-port", strconv.Itoa(testPort),
+					"--type", "NodePort", "--node-port", nodePort, "--protocol", "UDP")
+				err := virtctl()
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			It("Should be able to connect to the exposed service", func() {
+				By("Getting back the the service")
+				_, err := virtClient.CoreV1().Services(vm.Namespace).Get(serviceName, k8smetav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				By("Getting the node IP from all nodes")
+				nodes, err := virtClient.CoreV1().Nodes().List(k8smetav1.ListOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				Expect(nodes.Items).ToNot(BeEmpty())
+				for _, node := range nodes.Items {
+					Expect(node.Status.Addresses).ToNot(BeEmpty())
+					nodeIP := node.Status.Addresses[0].Address
+
+					By("Starting a pod which tries to reach the VM via NodePort")
+					job := tests.NewHelloWorldJobUDP(nodeIP, nodePort)
+					job, err = virtClient.CoreV1().Pods(vm.Namespace).Create(job)
+					Expect(err).ToNot(HaveOccurred())
+
+					By("Waiting for the pod to report a successful connection attempt")
+					getStatus := func() k8sv1.PodPhase {
+						pod, err := virtClient.CoreV1().Pods(job.Namespace).Get(job.Name, k8smetav1.GetOptions{})
+						Expect(err).ToNot(HaveOccurred())
+						return pod.Status.Phase
+					}
+					Eventually(getStatus, 60, 1).Should(Equal(k8sv1.PodSucceeded))
+				}
+			})
+		})
+	})
+
+	Context("Expose service on a VM replica set", func() {
+		const numberOfVMs = 2
+		By("Creating a VMRS object with 2 replicas")
+		template := tests.NewRandomVMWithEphemeralDiskAndUserdata(tests.RegistryDiskFor(tests.RegistryDiskCirros), "#!/bin/bash\necho 'hello'\n")
+		vmrs := tests.NewRandomReplicaSetFromVM(template, int32(numberOfVMs))
+		vmrs.Labels = map[string]string{"expose": "vmrs"}
+
+		It("Should start the VMRS and wait for all to be up", func() {
+			By("Start the replica set")
+			vmrs, err = virtClient.ReplicaSet(tests.NamespaceTestDefault).Create(vmrs)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Checking the number of ready replicas")
+			Eventually(func() int {
+				rs, err := virtClient.ReplicaSet(tests.NamespaceTestDefault).Get(vmrs.ObjectMeta.Name, k8smetav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				return int(rs.Status.ReadyReplicas)
+			}, 60*time.Second, 1*time.Second).Should(Equal(numberOfVMs))
+		})
+
+		It("Should add an 'hello world' server on each VM in the replica set", func() {
+			// TODO: add label to list options
```

Comment:
@rmohr I need help here. what can i pass into the ListOption object so it would find only the VMs that belong to the replica set?
would like to get rid of line 243

Reply:
The approach you have looks good. Why do you want the vms here?

Apart from that, you can pass in label selectors:

```golang
v12.ListOptions{LabelSelector: "key=val"})
```
This example would match all vms with the label `key` and value `val`.

Reply:
I need the VMs because I want to run the "hello world" server on them.
I tried providing the label selector as you suggested (with the label I set into the VMRS):
```v12.ListOptions{LabelSelector: "expose=vmrs"})```
However, the list return empty.
When I print the labels of the VMs in the VMRP it seems like the label that I set on the VMRS is not inherited by them, both ```vm.Labels``` and ```vm.ObjectMeta.Labels``` has the following value:
```map[kubevirt.io/nodeName:node01 name:replicaset99mjs]```
Is this by-design, or an issue?
anyway, will keep the tes tcode as-is, since I'm testing something else here.

Reply:
That is by design. The same like with every k8s controller. Controllers have their own labels which are distinct from the objects in their template section. You will have to add your `expose=vmrs` label on the label section on the vm, to also have it there.